python/ray/train/mosaic/mosaic_trainer.py

```diff
@@ -0,0 +1,168 @@
+from ray.air.config import DatasetConfig, RunConfig, ScalingConfig
+from ray.train.torch import TorchConfig, TorchTrainer
+from ray.train.trainer import GenDataset
+from ray.util import PublicAPI
+
+if TYPE_CHECKING:
+    from ray.data.preprocessor import Preprocessor
+
+
+@PublicAPI(stability="alpha")
+class MosaicTrainer(TorchTrainer):
+    """A Trainer for data parallel Mosaic Composers on PyTorch training.
+
+    This Trainer runs the ``composer.trainer.Trainer.fit()`` method on multiple
+    Ray Actors. The training is carried out in a distributed fashion through PyTorch
+    DDP. These actors already have the necessary torch process group already
+    configured for distributed PyTorch training.
+
+    The training function ran on every Actor will first run the
+    specified ``trainer_init_per_worker`` function to obtain an instantiated
+    ``composer.Trainer`` object. The ``trainer_init_per_worker`` function
+    will have access to preprocessed train and evaluation datasets.
+
+    Example:
+        .. code-block:: python
+
+    >>> def trainer_init_per_worker(**config):
+    ...     # prepare the model for distributed training and wrap with
+    ...     # ComposerClassifier for Composer Trainer compatibility
+    ...     model = torchvision.models.resnet18(num_classes=10)
+    ...     model = ComposerClassifier(ray.train.torch.prepare_model(model))
+    ...
+    ...     # prepare train dataloader
+    ...     batch_size_per_worker = BATCH_SIZE // session.get_world_size()
+    ...     train_dataloader = torch.utils.data.DataLoader(
+    ...         config.pop("train_dataset"),
+    ...         batch_size=batch_size_per_worker
+    ...     )
+    ...     train_dataloader = ray.train.torch.prepare_data_loader(train_dataloader)
+    ...
+    ...     # prepare optimizer
+    ...     optimizer = composer.optim.DecoupledSGDW(
+    ...         model.parameters(),
+    ...         lr=0.05,
+    ...         momentum=0.9,
+    ...         weight_decay=2.0e-3,
+    ...     )
+    ...
+    ...     return composer.trainer.Trainer(
+    ...         model=model,
+    ...         train_dataloader=train_dataloader,
+    ...         optimizers=optimizer,
+    ...         **config
+    ...     )
+    ...
+    >>> scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
+    >>> trainer_init_config = {
+    ...     "max_duration": "1ba",
+    ...     "train_dataset": train_dataset,
+    ...     "loggers": [InMemoryLogger()],
+    ...     "algorithms": [LabelSmoothing()],
+    ... }
+    ...
+    >>> trainer = MosaicTrainer(
+    ...     trainer_init_per_worker=trainer_init_per_worker,
+    ...     trainer_init_config=trainer_init_config,
+    ...     scaling_config=scaling_config,
+    ... )
+    ...
+    >>> trainer.fit()
+
+
+    Args:
+        trainer_init_per_worker: The function that returns an instantiated
+            ``composer.Trainer`` object and takes in an optional configuration
+            dictionary as an argument.
```

Comment:
Can we add more information on what this config dict is and why it is useful (explain that it's needed for hyparameter tuning so that different hyperparameter configurations can be passed in to the Trainer)?

Reply:
Further elaboration has been added